### ISSUE TITLE
Email forwarding

### ISSUE THREAD

**Issue by jimafisk** [axllent/mailpit#29]
I don't think this is the intention of the project, but wanted to throw an idea out there.

[Amazon SES](https://aws.amazon.com/ses/) is one of the more economic production SMTP services, but it doesn't offer an interface for seeing which emails are sent to who, at what time, with what contents - that's why projects like [SES Dashboard](https://sesdashboard.com/) exist.

Mailpit could provide all that context for a production app, but forward along the email to SES for deliverability. Too out of scope? Thanks!

**Comment by axllent** [comment]
@jimafisk Are you meaning that Mailpit would act as a form of transparent caching proxy for email, so it would route it through and store a copy?

**Comment by jimafisk** [comment]
Yes exactly. Not suggesting to completely change the direction of the project, trapping emails during development is very useful. Just wondering if it could provide additional capabilities for a prod environment to be an "invisible" layer between your app and SES to log the sent email info?

**Comment by axllent** [comment]
Thanks for the explanation. I'll give this some consideration (adding a setting of sorts for an SMTP relay) ~ though I'm a bit hesitant / cautious as something like this could then be used to illegally monitor routed email too. In saying that there are potentially already other existing ways to send duplicate copies of all mail to multiple hosts (I know you can do it by injecting `always_bcc=....` into postfix's config), which in turn would also solve your requirement.

**Comment by tonyswu** [comment]
For our usecase we only need one "release SMTP", so command line option would be sufficient.

**Comment by LiamKarlMitchell** [comment]
Would be nice to have a button in interface that could send email to another email address via some SMTP sender or web sending such as Mailgun/Sendgrid.
It would be useful to allow seeing how the email may appear in some specific email clients such as Outlook/Gmail/Apple and on Mobile device whilst still acting as a catch-all/trap for development servers.

**Comment by axllent** [comment]
I haven't started any coding for this yet, other than thinking about the most suitable implementation that will (hopefully) satisfy most use-cases. Currently my thinking is this:

- An runtime flag to set a single SMTP "relay server" configuration file. Whilst I'm sure some would like to relay via third-party APIs such as Mailgun or Sendgrid, the truth is there are too many different services to maintain, and most (if not all) support SMTP too. I also don't want to spend my time debugging different services.
- A runtime flag to automatically relay all received mails (ie: proxy). This would relay all emails to their original recipient(s) as they are received. There would be zero fault tolerance, so if the relay SMTP server was down or rejects the message for whatever reason, the relayed message would be lost (the error would be logged and the original mail would still be stored in Mailpit though). I do not want to implement a "postfix clone" to handle retries, rejections etc. Mailpit was never designed to be implemented in production as a "caching proxy", so if it is flooded with emails, then it will flood the relaying SMTP with emails too.
- If an SMTP relay configuration was provided during startup, the UI would have a button (per email) to release an email. This would default to the original To, Cc & Bcc values, but would allow you to change that before confirming sending.

A point worth noting is that relaying messages would alter somewhat the original mail headers (even if it is slightly), and potentially affect any present DKIM / SFP authentication depending on what relay SMTP is used and whether that aligns. There is no way to avoid this as that's exactly what DKIM / SFP authentication is designed to protect against.

**Comment by LiamKarlMitchell** [comment]
Found this from the past, thanks for your consideration.
> Mailpit does not have the ability to release emails (relay them to another SMTP). I have no plans to add this either.
https://github.com/axllent/mailpit/issues/15

Agree, SendGrid SMTP for example is well documented so totally fine to just configure the credentials in env vars probably.
https://docs.sendgrid.com/for-developers/sending-email/getting-started-smtp
```
User: YXBpa2V5 Base64 Encoded username of "apikey"
Pass: YourApiKey Base64 Encoded
Host: smtp.sendgrid.net
Port: SendGrid accepts unencrypted and TLS connections on ports 25, 587, & 2525. You can also connect via SSL on port 465. Many hosting providers and ISPs block port 25 as a default practice. (AWS does as well if I recall).
```

That makes sense on altering the headers slightly, a real server should be used anyway with something like mxtoolbox to check DKIM/SPF, Mostly curious about testing visually in various email clients and mobile clients manually. Something like https://www.litmus.com/email-testing/ peeks my curiosity but for most projects I only have a couple clients I'm worried about as sometimes styling/buttons come out wrong or other email clients muddle with or strip content, it's normally possible to switch them over to SendGrid temporarily just to check but one of the KEY reasons to use a service such as MailPit is that it traps the emails from development servers so they don't get sent to real addresses/inboxes by accident...

Great stuff thanks

**Comment by gosi-at** [comment]
> 

that would be greatly appreciated!

**Comment by patramsey** [comment]
@axllent I've put in some work on these features https://github.com/patramsey/mailpit/tree/release-messages in this branch. I can create a PR if it makes sense.

**Comment by axllent** [comment]
@patramsey Thank you. I have taken a quick look and it is actually quite similar to what I originally started doing several weeks ago (but gave up on temporarily). There is a lot more involved in this task to get it working, and also several related complications when it comes to auto-releasing messages - so I have much more to do, and several structural changes to make first. No need for a PR at this stage thanks, but please leave your branch as-is as it is handy for me to refer to when I do get around to restarting this. Email forwarding is definitely still on my radar, in fact it is the next big thing on my list, however I need to make some of the smaller changes first :+1:

**Comment by patramsey** [comment]
@axllent for sure, I'm happy to assist with the complications around auto-releasing messages if you'd like. 

I could separate the explicit release via the UI into a different branch from the auto release if that is useful. Maybe the changes you are envisioning don't impact that code path? The explicit release is the feature I'm most interested in and excited about. Our team will then be able to use Mailpit as a replacement for Mailhog!

**Comment by axllent** [comment]
Thank you for the offer @patramsey, but please hold off as there's a good chance I won't merge it, as it'll be more work to undo/redo things as it will be simply to write it the way I have in mind. I understand your urgency for wanting this, but you will just have to be a bit more patient.

**Comment by tyler36** [comment]
Would love to see this feature implemented. 

I recently created [ddev-mailpit](https://github.com/tyler36/ddev-mailpit) to intergrate Mailpit into a DDEV project. 
The feature discussed in this issue would help bring feature-parity with Mailhog (DDEV's default mail debugger) and maybe help replace it.

**Comment by axllent** [comment]
Despite my earlier comment about month-end release, I ended up completing this today (due to bad weather) and and have released it with v1.6.0.

This includes both the ability to "release" a message via the Web UI/API, as well as the original request "proxy" incoming mails automatically to a SMTP server. I have created a [wiki page](https://github.com/axllent/mailpit/wiki/SMTP-relay) to provide some additional documentation regarding the configuration file and the subtle differences between both options.

I'll close this issue given that there are now too many comments to constructively deal with feedback / issues with the new features. Any issues should be reported via a new Github issue, thanks. PS: I will still be away for a few days so responses will be delayed. Have fun!
